src/optix_coop_vec.cpp

```diff
@@ -0,0 +1,488 @@
+    } else {
+        declare_buffer("src", value);
+        copy_to_buffer("src", value);
+        put("        call (), _optix_reduce_sum_accumulate_ptr, (%type, %size, %dst, %offset, %src_p);\n");
+    }
+    put("    }\n");
+    if (use_mask)
+        fmt("\nl_$u_done:\n", v->reg_index);
+}
+
+void jitc_optix_render_coop_vec_outer_product_accum(const Variable *v, const Variable *target,
+                                                    const Variable *v0, const Variable *v1,
+                                                    const Variable *mask) {
+    uint32_t op_length = std::max(v0->array_length, v1->array_length),
+             reg_count = get_reg_count(op_length),
+             tsize = type_size[jitc_var(target->dep[3])->type];
+
+    const MatrixDescr *d = (const MatrixDescr *) v->data;
+
+    bool use_mask = !mask->is_literal() || mask->literal != 1;
+    if (use_mask)
+        fmt("    @!$v bra l_$u_done;\n", mask, v->reg_index);
+
+    fmt("    {   // coop_vec_outer_product_accum\n"
+        "        .reg.b32 %type_0, %type_1, %size_0, %size_1, %offset, %layout, %stride, %u;\n"
+        "        .reg.b64 %dst;\n"
+        "        mov.b32 %type_0, $u;\n"
+        "        mov.b32 %type_1, $u;\n"
+        "        mov.b32 %size_0, $u;\n"
+        "        mov.b32 %size_1, $u;\n"
+        "        mov.b32 %offset, $u;\n"
+        "        mov.b32 %layout, $u;\n"
+        "        mov.b32 %stride, $u;\n"
+        "        mov.b64 %dst, $v;\n",
+        jitc_optix_coop_vec_type_id((VarType) v0->type),
+        jitc_optix_coop_vec_type_id((VarType) v1->type),
+        v0->array_length,
+        v1->array_length,
+        d->offset * tsize,
+        jitc_optix_coop_vec_layout_id(d->layout),
+        d->stride * tsize,
+        target);
+
+    if (reg_count) {
+        fmt("        call (), _optix_outer_product_accumulate_$uxi32, (%type_0, "
+            "%size_0, %type_1, %size_1, %dst, %offset, %layout, %stride, ",
+            reg_count);
+        put_elem(v0, reg_count, true);
+        put_elem(v1, reg_count, false);
+        put(");\n");
+    } else {
+        declare_buffer("src_0", v0);
+        declare_buffer("src_1", v1);
+        copy_to_buffer("src_0", v0);
+        copy_to_buffer("src_1", v1);
+        put("        call (), _optix_outer_product_accumulate_ptr, (%type_0, "
+            "%size_0, %type_1, %size_1, %dst, %offset, %layout, %stride, "
+            "%src_0, %src_1);\n");
+    }
+    put("    }\n");
+    if (use_mask)
+        fmt("\nl_$u_done:\n", v->reg_index);
+}
+
+void jitc_optix_render_coop_vec(const Variable *v, const Variable *a0,
+                                const Variable *a1, const Variable *a2,
+                                const Variable *a3) {
+    uint32_t tsize = type_size[v->type],
+             length = v->array_length,
+             op_length = length;
+
+    if ((VarKind) v->kind == VarKind::CoopVecMatVec)
+        op_length = std::max(op_length, (uint32_t) a1->array_length);
```

Comment:
Explain what kind of broadcasting this allows?

Reply:
This is actually not related to broadcasting. I added this comment:

``` 
// For matrix multiplications, the size of the required intrinsic (e.g., 16
// vs 64 vs N) is related to the maximum cooperative vector size of the
// input and output.
```